Memory and style cleanup in StaticMemoryStrategy

jacquelinekay opened the pull request:
there were some mistakes made in #81 that weren't caught because Jenkins doesn't test the `StaticMemoryStrategy`. This PR makes some improvements to that and the executor:

Revert omission of `guard_condition_handles` in `StaticMemoryStrategy`; we still need that for the interrupt guard condition.
Introduce a `Subscription::return_message` class to ensure that memory allocated or borrowed by `create_message` is always returned.
Consolidate `get_next_ready_executable` functions.
Store `AnyExecutable` pool in `StaticMemoryStrategy` as shared pointers.
Use `new` instead of `malloc` in `StaticMemoryStrategy`.

jacquelinekay commented:
CI in progress:

http://ci.ros2.org/job/ros2_batch_ci_linux/177/
http://ci.ros2.org/job/ros2_batch_ci_osx/109/
http://ci.ros2.org/job/ros2_batch_ci_windows/149/

esteve commented:
+1

jacquelinekay commented:
OSX complained about mismatchAllocDealloc, rerunning with a fix:

http://ci.ros2.org/job/ros2_batch_ci_osx/110/

dirk-thomas commented on rclcpp/include/rclcpp/strategies/static_memory_strategy.hpp:
```diff
@@ -50,17 +51,23 @@ class StaticMemoryStrategy : public memory_strategy::MemoryStrategy
+    subscription_pool_ = new void *[bounds_.max_subscriptions_];
+    service_pool_ = new void *[bounds_.max_services_];
+    client_pool_ = new void *[bounds_.max_clients_];
+    guard_condition_pool_ = new void *[bounds_.max_guard_conditions_];
```
According to http://stackoverflow.com/questions/1087042/c-new-int0-will-it-allocate-memory I don't think the current approach is a good idea.

Instead I would suggest:

    if (bounds_.pool_size_) {
      memory_pool_ = new void *[bounds_.pool_size_];
    } else {
      memory_pool_ = 0;
    }

And then later for cleaning up:

    if (memory_pool_) {
      delete[] memory_pool_;
    }

tfoote commented:
+1

dirk-thomas commented on rclcpp/include/rclcpp/strategies/static_memory_strategy.hpp:
```diff
@@ -27,40 +27,59 @@ namespace memory_strategies
-    memset(client_pool_, 0, bounds_.max_clients_);
-    memset(executable_pool_, 0, bounds_.max_executables_);
+    if (bounds_.pool_size_) {
+      memory_pool_ = new void *[bounds_.pool_size_];
```
In my previous comment I was referring to the allocated void pointer when asking if they need to  be initialized with nullptr.

jacquelinekay replied:
yes, they should. I've restored the `memset` calls so that the memory that the `pool` objects is initialized to 0.

dirk-thomas commented on rclcpp/include/rclcpp/strategies/static_memory_strategy.hpp:
```diff
@@ -27,40 +27,64 @@ namespace memory_strategies
+  size_t pool_size_;
+
+  ObjectPoolBounds(size_t subs = 10, size_t services = 10, size_t clients = 10,
+    size_t executables = 1, size_t guard_conditions = 2, size_t pool = 1024)
```
I think this is a good example where the method chaining interface (which we added to the ROS 2 messages) is a better approach. The positional arguments are almost impossible to introspect for the user.

May be we should only provide a default constructor and setters?

wjwwood replied:
The only thing with setters is that you then have to consider what happens if someone calls a setter while a loan is out. Since we're using it we can assert we won't do that or we can throw an exception or something. But with the constructor arguments then there is no chance that this occurs since the destructor needs to be called before it can be constructed again.

Another option would be to have a factory object which can have setters for the settings and then when you're done with changing the settings you can call create which will return the object based on your factory's current settings. The object then has those settings until you delete it.

dirk-thomas replied:
The struct is being copied when passed into the memory strategy. Therefore calling any setter afterwards had no effect.

wjwwood replied:
I know, that's covered under the:

> Since we're using it we can assert we won't do that

wjwwood commented on rclcpp/include/rclcpp/strategies/static_memory_strategy.hpp:
```diff
@@ -27,40 +27,64 @@ namespace memory_strategies
+  size_t max_clients_;
+  size_t max_executables_;
+  size_t max_guard_conditions_;
+  size_t pool_size_;
```
Public members shouldn't have a trailing `_`.

I think @dirk-thomas is right, this class should be just like ROS messages where the attributes are public and there are setters for method chaining if desired and the constructor takes no arguments but sets reasonable defaults.

jacquelinekay commented:
I've implemented method chaining for the `ObjectPoolBounds` struct.

I've also changed the `StaticMemoryStrategy` example to use method chaining (and not to invoke the copy constructor in initialization of the shared pointer, which was stupid of me in the first place):

ros2/examples/pull/47

dirk-thomas commented:
+1

esteve commented:
+1 again

wjwwood commented:
lgtm, thanks for iterating on this.
